### PR TITLE
selector: Improve docs for implementors.

### PR DESCRIPTION
**Pull request by warpfork**
Writing selector implementations is hard.  The previous docs were very skeletal.  This should provide more knowledge about the phases and what's going on with the different functions a selector needs to implement.

### CODE REVIEW

File: traversal/selector/selector.go [hunk 1]
```diff
@@ -9,21 +9,102 @@ import (
...
-	Explore(ipld.Node, ipld.PathSegment) Selector // explore one step -- iteration comes from outside (either whole node, or by following suggestions of Interests).  returns nil if no interest.  you have to traverse to the next node yourself (the selector doesn't do it for you because you might be considering multiple selection reasons at the same time).
-	Decide(ipld.Node) bool
+	// Notes for you who implements a Selector:
+	// this type holds the state describing what we will do at one step in a traversal.
+	// The actual traversal stepping is applied *from the outside* (and this is implemented mostly in the `traversal` package;
+	// this type just gives it instructions on how to step.
+	// Each of the functions on this type should be pure; they can can read the Selector's fields, but should treat them as config, not as state -- the Selector should never mutate.
+	//
+	// The traversal process will ask things of a Selector in three phases,
+	// and control flow will bounce back and forth between the traversal logic --
```
Comment (willscott) [inline traversal/selector/selector.go:36]:
```suggestion
	// and control flow will bounce back and forth between traversal logic and selector evaluation --
```

File: traversal/selector/selector.go [hunk 2]
```diff
@@ -9,21 +9,102 @@ import (
...
+	// traversal owns the actual walking (and any data loading), and just briefly dips down into the Selector so it can answer questions:
+	//   T1. Traversal starts at some Node with some Selector.
+	//   S1. First, the traversal asks the Selector what its "interests" are.
+	//        This lets the Selector hint to the traversal process what it should load,
+	//        which can be important to performance purposes if not all of the next data elements are in memory already.
```
Comment (willscott) [inline traversal/selector/selector.go:41]:
```suggestion
	//        which can be important for performance if not all of the next data elements are in memory already.
```

File: traversal/selector/selector.go [hunk 3]
```diff
@@ -9,21 +9,102 @@ import (
...
+	//        (This may applicable especially for things like ADLs which contain large sharded data, for example.)
```
Comment (willscott) [inline traversal/selector/selector.go:42]:
```suggestion
	//        (This is applicable to ADLs which contain large sharded data, for example.)
```

File: traversal/selector/selector.go [hunk 4]
```diff
@@ -9,21 +9,102 @@ import (
...
+	// (In particular, a Selector implementation absolutely may **not** assume `Decide` will be called before `Interests`, and may **not** hold onto a Node statefully, etc.)
+	//
+	// Note that it's not until phase T2.a that the traversal actually loads child Nodes.
+	// This is interesting because it's *after* when the Selector is asked to `Explore` and yield a subsequent Selector to use on that upcoming Node.
+	//
+	// Can `Explore` and `Decide` do Link loading on their own?  Do they need to?
+	// Right now, no, they can't.  (Sort of.)  They don't have access to a LinkLoader; the traversal would have to give them one.
+	// This might be needed in the future, e.g. if the Selector has a Condition clause that requires looking deeper; so far, we don't have those features, so it hasn't been needed.
+	// The "sort of" is for ADLs.  ADLs that work with large sharded data sometimes hold onto their own LinkLoader and apply it transparently.
+	// In that case, of course, `Explore` and `Decide` can just interrogate the Node they've been given, and that make cause link loading.
```
Comment (willscott) [inline traversal/selector/selector.go:80]:
```suggestion
	// In that case, of course, `Explore` and `Decide` can just interrogate the Node they've been given, and that may cause link loading.
```

**Review by willscott — APPROVED** [review]

**Comment by warpfork** [comment]
Thanks for those studious catches on typos, @willscott !  :green_heart:

**Pull request merged**